gremlin/traversal/nexthop.go

```diff
@@ -0,0 +1,243 @@
+type NextHop struct {
+	IP      string
+	MAC     string
+	IfIndex int64
+}
+
+// NextHopTraversalExtension describes a new extension to enhance the topology
+type NextHopTraversalExtension struct {
+	NextHopToken traversal.Token
+}
+
+// NextHopGremlinTraversalStep nexthops step
+type NextHopGremlinTraversalStep struct {
+	context traversal.GremlinTraversalContext
+	ip      net.IP
+}
+
+// NewNextHopTraversalExtension returns a new graph traversal extension
+func NewNextHopTraversalExtension() *NextHopTraversalExtension {
+	return &NextHopTraversalExtension{
+		NextHopToken: traversalNextHopToken,
+	}
+}
+
+// ScanIdent returns an associated graph token
+func (e *NextHopTraversalExtension) ScanIdent(s string) (traversal.Token, bool) {
+	switch s {
+	case "NEXTHOP":
+		return e.NextHopToken, true
+	}
+	return traversal.IDENT, false
+}
+
+// ParseStep parses nexthops step
+func (e *NextHopTraversalExtension) ParseStep(t traversal.Token, p traversal.GremlinTraversalContext) (traversal.GremlinTraversalStep, error) {
+	switch t {
+	case e.NextHopToken:
+	default:
+		return nil, nil
+	}
+
+	if len(p.Params) != 1 {
+		return nil, fmt.Errorf("NextHop accepts one parameter : %v", p.Params)
+	}
+
+	ip, ok := p.Params[0].(string)
+	if !ok {
+		return nil, errors.New("NextHop parameter have to be a string")
+	}
+
+	if !common.IsIPv4(ip) {
+		return nil, errors.New("NextHop parameter have to be a valid IPv4 address")
+	}
+
+	return &NextHopGremlinTraversalStep{context: p, ip: net.ParseIP(ip)}, nil
+}
+
+func getNextHops(nodes []*graph.Node, ip net.IP) map[string]*NextHop {
+	nexthops := map[string]*NextHop{}
+
+nodeloop:
+	for _, node := range nodes {
+
+		neighbors, err := common.GetField(node.Metadata, "Neighbors")
+		if err == nil {
+			//First check given IP present in neighbors list.
+			hop := getNextHop(neighbors.(*netlink.Neighbors), ip)
+			if hop != nil {
+				nexthops[string(node.ID)] = hop
+				break nodeloop
+			}
+		}
+		tables, err := common.GetField(node.Metadata, "RoutingTables")
+		if err != nil {
+			continue
+		}
+		rts := tables.(*netlink.RoutingTables)
+		for _, t := range *rts {
+			var defaultRouteIP net.IP
+			var defaultIfIndex int64
+			for _, r := range t.Routes {
+				if r.Prefix != "" {
+					if common.IsIPinCIDR(ip, r.Prefix) {
+						nextIP := r.NextHops[0].IP
+						if nextIP == nil {
+							nexthops[string(node.ID)] = &NextHop{IfIndex: r.NextHops[0].IfIndex}
+							break nodeloop
+						}
+						if neighbors == nil {
+							nexthops[string(node.ID)] = &NextHop{IP: nextIP.String(), IfIndex: r.NextHops[0].IfIndex}
+							break nodeloop
+						}
+						hop := getNextHop(neighbors.(*netlink.Neighbors), nextIP)
+						if hop != nil {
+							nexthops[string(node.ID)] = hop
+							break nodeloop
+						}
+					}
+				} else {
+					if r.NextHops[0].IP == nil {
+						nexthops[string(node.ID)] = &NextHop{IfIndex: r.NextHops[0].IfIndex}
+						continue
+					}
+					if common.IsIPv4(r.NextHops[0].IP.String()) {
```

Comment:
this is to test the default IP is V4 because sometimes the default IP is V6.

Reply:
That's a good point. First It would be better to test the len. something like len(ip) == net.IPv6len. 

But it would be nice to address both ipv6 and ipv4, check the ip version of the given ip